src/write.rs

```diff
@@ -228,22 +241,24 @@ where
             first: true,
             second: true,
             taken: false,
+            yield_fn,
         }
     }
 
-    fn wait(&mut self, epochs: &mut MutexGuard<'_, slab::Slab<Arc<AtomicUsize>>>) {
+    fn wait(&mut self, epochs: &ListSnapshot) {
         let mut iter = 0;
         let mut starti = 0;
 
         #[cfg(test)]
         {
             self.is_waiting.store(true, Ordering::Relaxed);
         }
-        // we're over-estimating here, but slab doesn't expose its max index
-        self.last_epochs.resize(epochs.capacity(), 0);
+
+        // make sure we have enough space for all the epochs in the current snapshot
+        self.last_epochs.resize(epochs.iter().count(), 0);
```

Comment:
It's a little unfortunate to end up having to traverse the list twice. It'd be better if we could avoid that.